src/k8s/pkg/k8sd/features/calico/network.go

```diff
@@ -10,28 +10,53 @@ import (
 	"github.com/canonical/k8s/pkg/utils"
 )
 
+const (
+	enabledMsg          = "enabled"
+	disabledMsg         = "disabled"
+	deployFailedMsgTmpl = "Failed to deploy Calico, the error was: %v"
+	deleteFailedMsgTmpl = "Failed to delete Calico, the error was: %v"
+)
+
 // ApplyNetwork will deploy Calico when cfg.Enabled is true.
 // ApplyNetwork will remove Calico when cfg.Enabled is false.
-// ApplyNetwork returns an error if anything fails.
-func ApplyNetwork(ctx context.Context, snap snap.Snap, cfg types.Network, annotations types.Annotations) error {
+// ApplyNetwork will always return a FeatureStatus indicating the current status of the
+// deployment.
+// ApplyNetwork returns an error if anything fails. The error is also wrapped in the .Message field of the
+// returned FeatureStatus.
+func ApplyNetwork(ctx context.Context, snap snap.Snap, cfg types.Network, annotations types.Annotations) (types.FeatureStatus, error) {
+	status := types.FeatureStatus{
+		Version: calicoTag,
+		Enabled: cfg.GetEnabled(),
+	}
```

Comment:
avoid creating a `status` here that is mutated in place afterwards (as it can be easy to forget about fields in the future).

Reply:
I guess this and the following comments are a good amount of work (~0.5). will start changing them after the meetings.